parachain/pallets/erc20-app/src/payload.rs

```diff
@@ -28,10 +30,28 @@ pub struct InPayload<AccountId: codec::Decode> {
 
 impl<AccountId: codec::Decode> InPayload<AccountId> {
 
-	/// Decode payload into an Ethereum log event and then extract ABI-encoded data from log event
-	pub fn decode(payload: &[u8]) -> Result<Self, DecodeError> {
+	/// Decode payload into an Ethereum log event, check that it matches the verified receipt,
+	/// and then extract ABI-encoded data from log event
+	pub fn decode_verified(payload: &[u8], verification_output: &VerificationOutput) -> Result<Self, DecodeError> {
```

Comment:
this function looks the same across both apps - could it be in a shared library?

Reply:
Well I'm thinking 2 things:

1. Since Riz factored out most of the log decoding into the `TryFrom<Log>`  impl, we should consider moving the body of the reduced `decode_verified` into `Application::handle`. I'm thinking that's a more appropriate location since:
   * `Application::handle` doesn't do much else except call `decode_verified` and pass on the results.
   * This `payload.rs` module was mostly intended for decoding/unmarshalling, and "business rules" such as verification are maybe better suited for `Application::handle`.
2. I generally don't worry about factoring out duplicated code unless there are more than 2 duplications. As I've learnt from experience that sometimes those factorings, while nice and elegant, are often premature or wasted work in the end. That said, when our code is more stable and mature, we should definitely revisit and see how we can make it easier for users to develop new Substrate apps. - Just my 2cents.